File: sdk/typescript/packages/api/src/provider.ts
import type { Hook } from './hooks';

export interface ProviderCapabilities {
  track: boolean;
  identify: boolean;
  group: boolean;
  page: boolean;
  alias: boolean;
}

export interface ProviderMetadata {
  name: string;
  version: string;
  capabilities: ProviderCapabilities;
}

// MessageContext contains environment and device metadata attached to every message.
export interface MessageContext {
  ipAddress?: string;
  userAgent?: string;
  locale?: string;
  timezone?: string;
  library?: Record<string, unknown>;
  app?: Record<string, unknown>;
  device?: Record<string, unknown>;
  os?: Record<string, unknown>;
  network?: Record<string, unknown>;
  screen?: Record<string, unknown>;
  extra?: Record<string, unknown>;
}

export interface TrackMessage {
  messageId: string;
  timestamp: Date;
  eventName: string;
  properties: Record<string, unknown>;
  userId: string;
  anonymousId: string;
  context: MessageContext;
}

export interface IdentifyMessage {
  messageId: string;
  timestamp: Date;
  userId: string;
  anonymousId: string;
  traits: Record<string, unknown>;
  context: MessageContext;
}

export interface GroupMessage {
  messageId: string;
  timestamp: Date;
  userId: string;
  anonymousId: string;
  groupId: string;
  traits: Record<string, unknown>;
  context: MessageContext;
}

export interface PageMessage {
  messageId: string;
  timestamp: Date;
  userId: string;
  anonymousId: string;
  name: string;
  properties: Record<string, unknown>;
  context: MessageContext;
}

export interface AliasMessage {
  messageId: string;
  timestamp: Date;
  userId: string;
  previousId: string;
  context: MessageContext;
}

// ErrUnsupportedOperation is returned by provider methods that have no equivalent
// in the underlying vendor API, preventing silent data loss.
export const ErrUnsupportedOperation = new Error('unsupported provider operation');

// Provider is the interface every analytics destination adapter must satisfy.
// The single interface is intentional — providers that don't support a method
// should throw ErrUnsupportedOperation rather than silently no-oping.
export interface Provider {
  metadata(): ProviderMetadata;
  hooks(): Hook[];
  track(msg: TrackMessage): Promise<void>;
  identify(msg: IdentifyMessage): Promise<void>;
  group(msg: GroupMessage): Promise<void>;
  page(msg: PageMessage): Promise<void>;
  alias(msg: AliasMessage): Promise<void>;
  flush(): Promise<void>;
  shutdown(): Promise<void>;
}

File: sdk/kotlin/api/src/main/kotlin/io/eventspec/analytics/Provider.kt
package io.eventspec.analytics

import java.time.Instant

data class ProviderCapabilities(
    val track: Boolean = true,
    val identify: Boolean = true,
    val group: Boolean = true,
    val page: Boolean = true,
    val alias: Boolean = true,
)

data class ProviderMetadata(
    val name: String,
    val version: String,
    val capabilities: ProviderCapabilities = ProviderCapabilities(),
)

data class MessageContext(
    val ipAddress: String? = null,
    val userAgent: String? = null,
    val locale: String? = null,
    val timezone: String? = null,
    val library: Map<String, Any>? = null,
    val app: Map<String, Any>? = null,
    val device: Map<String, Any>? = null,
    val os: Map<String, Any>? = null,
    val network: Map<String, Any>? = null,
    val screen: Map<String, Any>? = null,
    val extra: Map<String, Any>? = null,
)

data class TrackMessage(
    val messageId: String,
    val timestamp: Instant,
    val eventName: String,
    val properties: Map<String, Any?>,
    val userId: String,
    val anonymousId: String,
    val context: MessageContext = MessageContext(),
)

data class IdentifyMessage(
    val messageId: String,
    val timestamp: Instant,
    val userId: String,
    val anonymousId: String,
    val traits: Map<String, Any?>,
    val context: MessageContext = MessageContext(),
)

data class GroupMessage(
    val messageId: String,
    val timestamp: Instant,
    val userId: String,
    val anonymousId: String,
    val groupId: String,
    val traits: Map<String, Any?>,
    val context: MessageContext = MessageContext(),
)

data class PageMessage(
    val messageId: String,
    val timestamp: Instant,
    val userId: String,
    val anonymousId: String,
    val name: String,
    val properties: Map<String, Any?>,
    val context: MessageContext = MessageContext(),
)

data class AliasMessage(
    val messageId: String,
    val timestamp: Instant,
    val userId: String,
    val previousId: String,
    val context: MessageContext = MessageContext(),
)

class UnsupportedOperationException(operation: String) :
    Exception("unsupported provider operation: $operation")

interface Provider {
  fun metadata(): ProviderMetadata

  fun hooks(): List<Hook> = emptyList()

  suspend fun track(msg: TrackMessage)

  suspend fun identify(msg: IdentifyMessage)

  suspend fun group(msg: GroupMessage)

  suspend fun page(msg: PageMessage)

  suspend fun alias(msg: AliasMessage)

  suspend fun flush()

  suspend fun shutdown()
}
